src/drawers/drawer.ts

```diff
@@ -76,7 +76,14 @@ export class Drawer {
    * @param{any[]} data The data to be drawn
    */
   private _bindSelectionData(data: any[]) {
-    let dataElements = this.selection().data(data);
+    // if the dataset has a key, use it when binding the data   
+    let dataElements: d3.selection.Update<any>;
+    if (this._dataset && this._dataset.key()) {
+      dataElements = this.selection().data(data, this._dataset.key());
+    } else {
+      dataElements = this.selection().data(data);
+    }
```

Comment:
I agree with you that it feels hacky, but I don't want to break that.
Stepping back a bit, the goal here ultimately is to support animations that can enrich the data visualisation by allowing the user to see how data points change from dataset to dataset, and to be made aware of points that enter the dataset, and points that leave. 
This [quote] (http://bost.ocks.org/mike/constancy/) from Mike Bostok sums this up I think:
_Above all, animation should be meaningful. While it may be visually impressive for bars to fly around the screen during transitions, animation should only be used when it enhances understanding._

I have banged on about this at some length in this issue:  #2512 
To get to that point in plottable, there are a couple of things that need to happen, and this PR is the first of them:
1) we need a way to specify a key function that allows a meaningful connection to be made between data points in different datasets (This PR does that much)
2) we need to support distinct animations on entering and exiting points - the [enter/update/exit] (https://gist.github.com/mbostock/3808218) is hidden inside Drawer - I believe that it needs to be exposed to the Animator. 
3) Once there is a key to define the relationship between elements, and the Animator can manipulate each of the update, enter and exit selections, we need new versatile concrete animators to take advantage of that. 

If this paradigm is in place, then resetting the data to identical data should not cause any animation. If you want the animation as it is currently implemented, then you could explicitly use the ```noConstancy``` key function to ensure there is never any join between the incoming and outgoing datasets - this means that every datum will be included in the enter() selection, and handled by the Animator accordingly.

But the current ```animateOnNextRender``` implementations I would argue are incompatible with this "object constancy" approach to animation, so they probably should go.

Now step 2) above is probably more radical and I'm fleshing this out now - but I'd need guidance from you about changing (breaking?) the IAnimator interface,and I can appreciate you may want to take a careful look at that before heading down this path.
But if we put back the piece in ScatterPlot, the other changes would not affect any existing code - data binding would use the default key ```useIndex``` which is the d3 default anyway.